### PR TITLE
Added Cosine Similarity metric

### LINKED ISSUE #75: [New metric] CosineSimilarity
## 🚀 Feature

Add Cosine similarity metric
https://en.wikipedia.org/wiki/Cosine_similarity
Measures the similarity between two feature vectors by calculating the angle between them.

### PR DESCRIPTION
**Pull request by GiannisVagionakis**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Resolves #75

I am co-worker of @N-georgakopoulos (issue assigned to him: https://github.com/PyTorchLightning/metrics/issues/75) and so I made this PR with good faith that it solves the issue sufficiently as regards the basic implementation, being more than happy to follow any advice or directive from someone more exprerienced

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
@GiannisVagionakis great additions, mind check
- [ ] add docs
- [ ] add tests

File: torchmetrics/classification/cosine_similarity.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        self.add_state("correct", default=tensor(0), dist_reduce_fx="sum")
+        self.add_state("total", default=tensor(0), dist_reduce_fx="sum")
+        self.reduction = reduction
+
+        if not 0 < threshold < 1:
+            raise ValueError("The `threshold` should lie in the (0,1) interval.")
+        self.threshold = threshold
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:83]:
```suggestion
```

File: torchmetrics/classification/cosine_similarity.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+           >>> preds = torch.tensor([[0, 1], [0, 1]])
+           >>> cosine_similarity = CosineSimilarity()
+           >>> cosine_similarity.update(preds, target, 'none')
+           >>> cosine_similarity.compute()
+           tensor([1.0000, 1.0000, 1.0000])
+    """
+
+    def __init__(
+        self,
+        threshold: float = 0.5,
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:63]:
```suggestion
```

File: torchmetrics/classification/cosine_similarity.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+        reduction: str = 'sum'
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:68]: let this be the first parameter

File: torchmetrics/classification/cosine_similarity.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        self.add_state("correct", default=tensor(0), dist_reduce_fx="sum")
+        self.add_state("total", default=tensor(0), dist_reduce_fx="sum")
+        self.reduction = reduction
+
+        if not 0 < threshold < 1:
+            raise ValueError("The `threshold` should lie in the (0,1) interval.")
+        self.threshold = threshold
+
+    def update(self, preds: Tensor, target: Tensor, reduction="sum"):
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:85]:
```suggestion
    def update(self, preds: Tensor, target: Tensor):
```

File: torchmetrics/classification/cosine_similarity.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        Update state with predictions and targets. See :ref:`references/modules:input types` for more information
+        on input types.
+        Args:
+            preds: Predictions from model (probabilities, logits or labels)
+            target: Ground truth labels
+            reduction : how to reduce over the batch dimension using sum, mean or
+                        taking the individual scores
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:93]: copy this text to the docstring of the init method
```suggestion

```

File: torchmetrics/classification/cosine_similarity.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        correct, total = _cosine_similarity_update(preds, target)
+
+        self.correct += correct
+        self.total += total
+        self.reduction = reduction
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:99]:
```suggestion

```

File: torchmetrics/classification/cosine_similarity.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+               Specify the process group on which synchronization is called.
+               default: ``None`` (which selects the entire world)
+           dist_sync_fn:
+               Callback that performs the allgather operation on the metric state. When ``None``, DDP
+               will be used to perform the all gather.
+           reduction: The method of reducing along the batch dimension using sum, mean or
+                        taking the individual scores
+       Raises:
+           ValueError:
+               If ``threshold`` is not between ``0`` and ``1``.
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:50]:
```suggestion
```

File: torchmetrics/classification/cosine_similarity.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+class CosineSimilarty(Metric):
+    r"""
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+       Accepts all input types listed in :ref:`references/modules:input types`.
+       Args:
+           threshold:
+               Threshold for transforming probability or logit predictions to binary (0,1) predictions, in the case
+               of binary or multi-label inputs. Default value of 0.5 corresponds to input being probabilities.
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:34]: should be exchanged with a docstring of the `reduction` arg
```suggestion

```

**Review by SkafteNicki — COMMENTED** [review]
Thanks for wanting to contribute. Most comments are about the `threshold` parameter which seems to be a copy-paste error from another metric :]

**Comment by GiannisVagionakis** [comment]
Hi, writting some documentation now, with respect to updating the pull request by the weekend! I would like however to understand a little bit more about input types? Is the implementation example i've put in the comments correct?

File: torchmetrics/classification/cosine_similarity.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+)
+from torchmetrics.metric import Metric
+
+
+class CosineSimilarity(Metric):
+    r"""
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+       Accepts all input types listed in :ref:`references/modules:input types`.
+       Args:
```
Comment (Borda) [inline torchmetrics/classification/cosine_similarity.py:31]:
```suggestion

       Args:
```

File: torchmetrics/classification/cosine_similarity.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+           dist_sync_on_step:
+               Synchronize metric state across processes at each ``forward()``
+               before returning the value at the step.
+           process_group:
+               Specify the process group on which synchronization is called.
+               default: ``None`` (which selects the entire world)
+           dist_sync_fn:
+               Callback that performs the allgather operation on the metric state. When ``None``, DDP
+               will be used to perform the all gather.
+       Example:
```
Comment (Borda) [inline torchmetrics/classification/cosine_similarity.py:45]:
```suggestion

       Example:
```

File: torchmetrics/classification/cosine_similarity.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+        self.add_state("correct", default=tensor(0), dist_reduce_fx="sum")
+        self.add_state("total", default=tensor(0), dist_reduce_fx="sum")
+        self.reduction = reduction
+
+    def update(self, preds: Tensor, target: Tensor):
+        """
+        Update state with predictions and targets. See :ref:`references/modules:input types` for more information
+        on input types.
+        Args:
```
Comment (Borda) [inline torchmetrics/classification/cosine_similarity.py:78]:
```suggestion

        Args:
```

File: torchmetrics/functional/classification/cosine_similarity.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    reduction_mapping = {"sum": torch.sum, "mean": torch.mean, "none": lambda x: x}
+    return reduction_mapping[reduction](similarity)
+
+
+def cosine_similarity(preds: Tensor, target: Tensor, reduction='sum') -> Tensor:
+    r"""
+        Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+        Accepts all input types listed in :ref:`references/modules:input types`.
+        Args:
```
Comment (Borda) [inline torchmetrics/functional/classification/cosine_similarity.py:44]:
```suggestion

        Args:
```

**Comment by Borda** [comment]
> Hi, writting some documentation now, with respect to updating the pull request by the weekend!

if so we may include it in net release :]

> I would like however to understand a little bit more about input types? Is the implementation example i've put in the comments correct?

@SkafteNicki mind advice, or ping him on slack :]

File: docs/source/references/modules.rst [hunk 13]
```diff
@@ -214,6 +214,12 @@ ConfusionMatrix
 .. autoclass:: torchmetrics.ConfusionMatrix
     :noindex:
 
+CosineSimilarity
+~~~~~~~~~~~~~~~
```
Comment (SkafteNicki) [inline docs/source/references/modules.rst:218]:
```suggestion
~~~~~~~~~~~~~~~~
```

File: torchmetrics/classification/cosine_similarity.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    _cosine_similarity_compute,
+    _cosine_similarity_update,
+)
+from torchmetrics.metric import Metric
+
+
+class CosineSimilarity(Metric):
+    r"""
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:29]: just some math :]
```suggestion
        between targets and predictions:
        
        .. math::
            cos_{sim}(x,y) = \frac{x \cdot y}{||x|| \cdot ||y|| = \frac{\sum_{i=1}^n x_i y_i}{\sqrt{\sum_{i=1}^n x_i^2} \sqrt{\sum_{i=1}^n y_i^2}}
            
        where :math:`y` is a tensor of target values, and :math:`x` is a tensor of predictions.
```

File: torchmetrics/classification/cosine_similarity.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+       Accepts all input types listed in :ref:`references/modules:input types`.
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:30]: This is not correct as it refers to the input types for classification problems.
Instead is should be something similar to 
https://github.com/PyTorchLightning/metrics/blob/652d3f804cad08660842fad440cdead2e986059d/torchmetrics/regression/pearson.py#L36-L39

File: torchmetrics/functional/classification/cosine_similarity.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+from torch import Tensor
+
+
+def _cosine_similarity_update(
+    preds: Tensor,
+    target: Tensor,
+) -> Tuple[Tensor, int]:
+    preds.float()
+    target.float()
+
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/cosine_similarity.py:26]: call this function here to check that they are same shape
https://github.com/PyTorchLightning/metrics/blob/21ce4f64c9446fe5b98d96ea1dab65caa81441ea/torchmetrics/utilities/checks.py#L23-L26

File: torchmetrics/classification/cosine_similarity.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+               will be used to perform the all gather.
+
+       Example:
+           >>> from torchmetrics import CosineSimilarity
+           >>> target = torch.tensor([[0, 1], [1, 1]])
+           >>> preds = torch.tensor([[0, 1], [0, 1]])
+           >>> cosine_similarity = CosineSimilarity()
+           >>> cosine_similarity.update(preds, target)
+           >>> cosine_similarity.compute()
+           tensor([1.0000, 1.0000, 1.0000])
```
Comment (SkafteNicki) [inline torchmetrics/classification/cosine_similarity.py:54]: i am pretty sure the output should be a scalar

**Review by SkafteNicki — COMMENTED** [review]
Hi @GiannisVagionakis,
Great work! In my opinion the cosine similarity is not a classification metric but a regression metrics (because the target tensor can also be float values). Therefore please move the files to the appropriate folders.

**Comment by GiannisVagionakis** [comment]
> Hi @GiannisVagionakis,
> Great work! In my opinion the cosine similarity is not a classification metric but a regression metrics (because the target tensor can also be float values). Therefore please move the files to the appropriate folders. 

I moved the files to the right folders, and changed the docs too

> i am pretty sure the output should be a scalar

i was actually trying to refer to the case where the reduction arg equals "none", returning the individual scores. 
I changed the docstrings in both files according to yous suggestions

**Review by Borda — APPROVED** [review]

File: torchmetrics/functional/regression/cosine_similarity.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+from torchmetrics.utilities.checks import _check_same_shape
+
+
+def _cosine_similarity_update(
+    preds: Tensor,
+    target: Tensor,
+) -> Tuple[Tensor, Tensor]:
+    _check_same_shape(preds, target)
+    preds = Tensor.float(preds)
+    target = Tensor.float(target)
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/cosine_similarity.py:28]: just preference :]
```suggestion
    preds = preds.float()
    target = target.float()
```

File: torchmetrics/functional/regression/cosine_similarity.py [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+                        taking the individual scores
+
+        Example:
+            >>> from torchmetrics.functional.regression import cosine_similarity
+            >>> target = torch.tensor([[1, 2, 3, 4],
+            ...                        [1, 2, 3, 4]])
+            >>> preds = torch.tensor([[1, 2, 3, 4],
+            ...                       [1, 2, 3, 4]])
+            >>> cosine_similarity(preds, target, 'none')
+            tensor([1.0000, 1.0000])
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/cosine_similarity.py:64]: just to not have the precise example twice :]
```suggestion
            >>> preds = torch.tensor([[1, 2, 3, 4],
            ...                       [-1, -2, -3, -4]])
            >>> cosine_similarity(preds, target, 'none')
            tensor([1.0000, -1.0000])
```

File: torchmetrics/functional/regression/cosine_similarity.py [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+        Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+        Forward accepts
+
+        - ``preds`` (float tensor): ``(N,)``
+        - ``target``(float tensor): ``(N,)``
+
+        Args:
+            preds: Predictions from model (probabilities, logits or labels)
+            target: Ground truth
```
Comment (SkafteNicki) [inline torchmetrics/functional/regression/cosine_similarity.py:53]: functional has no forward :]
```suggestion

        Args:
            preds: Predicted tensor with shape ``(N,d)``
            target: Ground truth tensor with shape ``(N,d)``
```

File: torchmetrics/regression/cosine_similarity.py [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+from torchmetrics.metric import Metric
+from torchmetrics.utilities.data import dim_zero_cat
+
+
+class CosineSimilarity(Metric):
+    r"""
+       Computes the `Cosine Similarity <https://en.wikipedia.org/wiki/Cosine_similarity>`_
+        between targets and predictions:
+
+        .. math::
```
Comment (SkafteNicki) [inline torchmetrics/regression/cosine_similarity.py:29]: could you copy the math to the functional docstring

File: torchmetrics/regression/cosine_similarity.py [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+            cos_{sim}(x,y) = \frac{x \cdot y}{||x|| \cdot ||y|| = \frac{\sum_{i=1}^n x_i y_i}{\sqrt{\sum_{i=1}^n x_i^2}
+             \sqrt{\sum_{i=1}^n y_i^2}}
+
+        where :math:`y` is a tensor of target values, and :math:`x` is a tensor of predictions.
+
+        Forward accepts
+
+        - ``preds`` (float tensor): ``(N,)``
+        - ``target``(float tensor): ``(N,)``
```
Comment (SkafteNicki) [inline torchmetrics/regression/cosine_similarity.py:38]:
```suggestion
        - ``preds`` (float tensor): ``(N,d)``
        - ``target``(float tensor): ``(N,d)``
```

File: torchmetrics/regression/cosine_similarity.py [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+        """
+        Update state with predictions and targets. Forward accepts
+
+        - ``preds`` (float tensor): ``(N,)``
+        - ``target``(float tensor): ``(N,)``
+
+        Args:
+            preds: Predictions from model (probabilities, logits or labels)
+            target: Ground truth labels
+        """
```
Comment (SkafteNicki) [inline torchmetrics/regression/cosine_similarity.py:93]:
```suggestion
        """
        Update metric states with predictions and targets.
        Args:
            preds: Predicted tensor with shape ``(N,d)``
            target: Ground truth tensor with shape ``(N,d)``
        """
```

**Review by SkafteNicki — COMMENTED** [review]
Great job. A bit a comments to the docstring of metric :]

File: torchmetrics/regression/cosine_similarity.py (hunk 23 above)
Comment (GiannisVagionakis) [reply torchmetrics/regression/cosine_similarity.py:93]: @SkafteNicki alright, working on it! Is there anything more needed to be done? for example I saw a failed check on tests with ddp, which I could not run locally as I'm working on a Windows machine

**Comment by GiannisVagionakis** [comment]
@SkafteNicki alright, working on it! Is there anything more needed to be done? for example I saw a failed check on tests with ddp, which I could not run locally as I'm working on a Windows machine (same applies to make-docs)

**Comment by SkafteNicki** [comment]
@GiannisVagionakis I will look into the failing test soon :]

**Comment by SkafteNicki** [comment]
Hi @GiannisVagionakis,
So I (hopefully) fixed the doctest and have currently removed `none` case from testing as it is making the other test fails.
This has nothing to do with the calculation not being correctly for `none` but has something to do with the underlying testing interface giving data to the comparing function in a different order than your implementation.
I send a follow up PR that corrects this and re-adds the testing of the `none` case.

**Review by SkafteNicki — APPROVED** [review]

**Comment by GiannisVagionakis** [comment]
@SkafteNicki  @Borda thanks for all the helpfunessl, patience and approachability! Cheers! :]
